### PR TITLE
Fix escaping brackets in setenv

### PR DESCRIPTION
**Pull request by jayvdb**
Fixs regression in v3.20.0 that caused escaped curly braces in setenv to break usage of the variable elsewhere in tox.ini.

Fixes https://github.com/tox-dev/tox/issues/1690

## Contribution checklist:

(also see [CONTRIBUTING.rst](../tree/master/CONTRIBUTING.rst) for details)

- [x] wrote descriptive pull request text
- [x] added/updated test(s)
- [ ] updated/extended the documentation
- [x] added relevant [issue keyword](https://help.github.com/articles/closing-issues-using-keywords/)
      in message body
- [x] added news fragment in [changelog folder](../tree/master/docs/changelog)
  * fragment name: `<issue number>.<type>.rst` for example (588.bugfix.rst)
  * `<type>` is must be one of `bugfix`, `feature`, `deprecation`,`breaking`, `doc`, `misc`
  * if PR has no issue: consider creating one first or change it to the PR number after creating the PR
  * "sign" fragment with "by :user:`<your username>`"
  * please use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files - by :user:`superuser`."
  * also see [examples](../tree/master/docs/changelog)
- [x] added yourself to `CONTRIBUTORS` (preserving alphabetical order)

### CODE REVIEW

**Comment by jayvdb** [comment]
The macOS CI error appears to be unrelated.

File: src/tox/config/__init__.py [hunk 1]
```diff
@@ -410,6 +408,19 @@ def __setitem__(self, name, value):
         self.definitions[name] = value
         self.resolved[name] = value
 
+    def items(self):
+        return [(name, self.__getitem__(name)) for name in self.definitions]
```
Comment (asottile) [inline src/tox/config/__init__.py:412]:
```suggestion
        return ((name, self[name]) for name in self.definitions)
```

1. more closely match the python3 `.items()` api
2. don't call double-double-under methods

File: src/tox/config/__init__.py [hunk 2]
```diff
@@ -410,6 +408,19 @@ def __setitem__(self, name, value):
...
+
+    def export(self):
+        # post-process items to avoid internal syntax/semantics
+        # such as {} being escaped using \{\}, suitable for use with
+        # os.environ .
+        return {
+            name: re.sub(r"\\({|})", r"\1", value)
```
Comment (asottile) [inline src/tox/config/__init__.py:419]: regex is overkill here, I'd just use `.replace` -- it looks like the original code did this why did you change it? (the original code is much more readable imo)

File: tests/unit/test_venv.py [hunk 3]
```diff
@@ -799,6 +803,10 @@ def test_env_variables_added_to_pcall(tmpdir, mocksession, newconfig, monkeypatc
         assert env is not None
         assert "ENV_VAR" in env
         assert env["ENV_VAR"] == "value"
+        assert env["ESCAPED_VAR"] == "{value}"
+        assert env["ESCAPED_VAR2"] == r"\{value\}"
+        assert env["COLON"] == ";" if sys.platform == "win32" else ":"
+        assert env["TTY_VAR"] == "OFF_VALUE"
```
Comment (asottile) [inline tests/unit/test_venv.py:809]: these seem unrelated to the patch

**Review by asottile — COMMENTED** [review]
it's unclear to me what this is actually fixing, why would you still want the backslashes in the variable?

can you show a concrete usecase?

File: tests/unit/test_venv.py (hunk 3 above)
Comment (jayvdb) [reply tests/unit/test_venv.py:809]: these are testing other parts of substitution internals which are not tested in this context.  If `export()` is to work correctly with a stable API, all of the substitution internal values need to be converted into externally understandable values.

For the record, using `{posargs}` in setenv is another case of special handling in the substitution system, and it fails strangely in setenv.

And another internal voodoo not tested is `file|` - I havent tried that yet.

File: src/tox/config/__init__.py (hunk 2 above)
Comment (jayvdb) [reply src/tox/config/__init__.py:419]: I didnt pay attention to the original code, as it was in a very different area of the code exercised by different tests.  I'm happy to use `.replace`.

**Comment by jayvdb** [comment]
> why would you still want the backslashes in the variable?

backslashes are how tox escapes `{` so that `{` can be a literal in a variable value.
This is in the docs at https://tox.readthedocs.io/en/latest/config.html#substitutions

If you are asking about literal backslashes in the variable value, there are many reasons people might want them.
https://www.google.com/search?q=backslashes+in+environment+variables

File: src/tox/config/__init__.py (hunk 2 above)
Comment (gaborbernat) [reply src/tox/config/__init__.py:419]: Yeah please do.

**Review by gaborbernat — CHANGES_REQUESTED** [review]

File: tests/unit/test_venv.py (hunk 3 above)
Comment (jayvdb) [reply tests/unit/test_venv.py:809]: I've tested with `file|` - no problems here, but https://github.com/tox-dev/tox/issues/1694 raised for other issues.

I also retested with `{posargs}`, `{packages}` and `{opts}` and very odd behaviour raised in https://github.com/tox-dev/tox/issues/1695

File: src/tox/config/__init__.py (hunk 2 above)
Comment (jayvdb) [reply src/tox/config/__init__.py]: This was done.  Please review again when you have time, as this PR merge conflicts with other PRs, and is also a prerequisite for testing in other PRs using brackets in settings and their effect on the runtime environment variables present while running `commands`.

**Review by gaborbernat — APPROVED** [review]
